### PR TITLE
Add checkpoint verification

### PR DESCRIPTION
**Pull request by loosebazooka**
negative test cases are here: https://github.com/sigstore/sigstore-conformance/pull/139

it's always a bit weird to mirror those test cases in sigstore-java, but I guess I can?

### CODE REVIEW

**Review by haydentherapper — APPROVED** [review]
Looks great!

File: sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java [hunk 1]
```diff
@@ -134,6 +138,43 @@ private void verifyInclusionProof(RekorEntry entry) throws RekorVerificationExce
...
+    byte[] checkpointRootHash = Base64.getDecoder().decode(checkpoint.getBase64Hash());
+
+    if (!Arrays.equals(inclusionRootHash, checkpointRootHash)) {
+      throw new RekorVerificationException(
+          "Checkpoint root hash does not match root hash provided in inclusion proof");
+    }
+    var keyHash = Hashing.sha256().hashBytes(tlog.getPublicKey().getRawBytes()).asBytes();
+    // checkpoint 0 is always the log, not any of the cross signing verifiers/monitors
+    var sig = checkpoint.getSignatures().get(0);
+    for (int i = 0; i < 4; i++) {
```
Comment (bobcallaway) [inline sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:161]: what is the `4` for?

Comment (haydentherapper) [reply sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:161]: Key ID is 4 bytes, https://github.com/sigstore/rekor/blob/main/pkg/util/signed_note.go#L140, truncated public key hash

Comment (haydentherapper) [reply sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:161]: Would be good to have a link to that or https://github.com/C2SP/C2SP/blob/main/signed-note.md#signatures

Comment (bobcallaway) [reply sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:161]: that should be added in a comment at least IMO

Comment (loosebazooka) [reply sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:161]: It's defined in the spec and the spec is linked in the checkpoint type

Comment (loosebazooka) [reply sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:161]: https://github.com/sigstore/sigstore-java/blob/main/sigstore-java/src/main/java/dev/sigstore/rekor/client/RekorEntry.java#L106

**Pull request merged**